src/integrator_interface.jl

```diff
@@ -37,34 +37,34 @@ end
 Calculate next step of `integrator`.
 """
 function perform_step!(integrator::DDEIntegrator)
-    # update previous time to extrapolate from current interval
-    integrator.tprev = integrator.t
-
     # update ODE integrator
     integrator.integrator.uprev = integrator.uprev
-    integrator.integrator.tprev = integrator.tprev
-    integrator.integrator.fsalfirst = integrator.fsalfirst
+    integrator.integrator.tprev = integrator.t # to extrapolate from current interval
     integrator.integrator.t = integrator.t
     integrator.integrator.dt = integrator.dt
 
-    # if dt is greater than the minimal lag, then it's explicit so use fixed-point iteration
-    if integrator.dt > minimum(integrator.prob.lags)
+    # separate u to avoid influence of calculations on interpolation and to calculate
+    # residuals
+    if typeof(integrator.u) <: AbstractArray
+        recursivecopy!(integrator.u_cache, integrator.u)
+    else
+        integrator.u_cache = integrator.u
+    end
+    integrator.integrator.u = integrator.u_cache
 
-        # save these values to correct the extrapolation after the last iteration
-        tprev_cache = integrator.tprev
-        t_cache = integrator.t # same as tprev_cache?
-        uprev_cache = integrator.uprev
+    # separate k to avoid influence of calculations on interpolation
+    # constant caches that contain fsalfirst must be updated for interpolation
+    # has to be changed all such caches - move to OrdinaryDiffEq?
+    if typeof(integrator.cache) <: Union{BS3ConstantCache}
```

Comment:
To update `k[1]`. For some caches like `BS3ConstantCache` it contains `fsalfirst` - however, at this point of the algorithm its an old, outdated value of `fsalfirst`, and hence also any interpolation that uses `k[1]` is wrong. There are no problems with `BS3Cache` since it only points to `fsalfirst` and hence the value of `k[1]` always equals `fsalfirst`.

Of course, one has to add all affected caches (there are many more besides `BS3ConstantCache` - all not-inplace caches that contain `fsalfirst` and use it for interpolation) to this `Union`. I just wanted to wait til I know whether you would like to apply this update of `k` here or rather in `OrdinaryDiffEq.jl` when `fsalfirst` is updated.